### DISCUSSION TITLE
Automating Class Attribute Documentation Generation

### DISCUSSION THREAD

**Discussion by nh916** [squidfunk/mkdocs-material#5930] [Q&A]
I'm currently working on documenting my Python package, and I have a question about documenting class attributes. For each of my classes, I have created a dataclass that holds all the class attributes throughout the class.

```python
class User:
    @dataclass(frozen=True)
    class Attributes(PrimaryAttributes):
        email: Optional[str] = ""
        model_version: str = ""
        username: str = ""
...
```

I want to include a table at the top of the documentation for each class that showcases these attributes along with their types, examples, descriptions, and whether they are required. Currently, I'm manually creating this table for each class, and it's proving to be time-consuming, hard to maintain, and error-prone. I'm looking for an alternative approach to streamline this process.

In essence, I'm seeking a solution that can automate the generation of a table like the one below:
![image](https://github.com/squidfunk/mkdocs-material/assets/26590757/bb0fb1c5-154b-4cd9-a463-cae7173bc5f4)

```md
| attribute  | type      | example                    | description                                | required |
|------------|-----------|----------------------------|--------------------------------------------|----------|
| url        | str       |                            | unique ID of the node                      | True     |
| username   | str       | "john_doe"                 | User’s name                                | True     |
| email      | str       | "user@website.com"         | email of the user                          | True     |
| updated_at | datetime* | 2023-03-06 18:45:23.450248 | last date the node was modified (UTC time) | True     |
| created_at | datetime* | 2023-03-06 18:45:23.450248 | date it was created (UTC time)             | True     |
````

If anyone has experience or insights into automating this process or using tools to achieve this kind of documentation generation, I'd greatly appreciate your guidance.

Please let me know if my explanation needs further clarification. Thank you!

**Comment by squidfunk** [comment]
Have you looked at [mkdocstrings](https://github.com/mkdocstrings/mkdocstrings) by @pawamoy?
